src/typechecker.erl

```diff
@@ -2540,47 +2553,45 @@ do_type_check_expr_in(Env, ResTy, Expr = {'fun', P, {function, M, F, A}}) ->
             end;
         _ -> % We don't have enough information to check the type.
-            {#{}, constraints:empty()}
+            {Env, constraints:empty()}
     end;
 do_type_check_expr_in(Env, Ty, {named_fun, _, FunName, Clauses} = Fun) ->
-    NewEnv = Env#env{ venv = add_var_binds(#{ FunName => Ty }, Env#env.venv, Env) },
+    Env1 = add_var_binds(Env#env{venv = #{ FunName => Ty }}, Env, Env),
     case expect_fun_type(Env, Ty) of
         any ->
-            {#{ FunName => Ty }, constraints:empty()};
+            {Env#env{venv = #{ FunName => Ty }}, constraints:empty()};
         {fun_ty, ArgsTy, ResTy, Cs1} ->
-            {VB, Cs2} = check_clauses(NewEnv, ArgsTy, ResTy, Clauses, bind_vars),
-            {VB, constraints:combine(Cs1, Cs2)};
+            {Env2, Cs2} = check_clauses(Env1, ArgsTy, ResTy, Clauses, bind_vars),
+            {Env2, constraints:combine(Cs1, Cs2)};
         {fun_ty_any_args, ResTy, Cs1} ->
-            {VB, Cs2} = check_clauses(NewEnv, any, ResTy, Clauses, bind_vars),
-            {VB, constraints:combine(Cs1, Cs2)};
+            {Env2, Cs2} = check_clauses(Env1, any, ResTy, Clauses, bind_vars),
+            {Env2, constraints:combine(Cs1, Cs2)};
         %% TODO: Can this case actually happen?
         {fun_ty_intersection, Tyss, Cs1} ->
-            {VB, Cs2} = check_clauses_intersect(NewEnv, Tyss, Clauses),
-            {VB, constraints:combine(Cs1, Cs2)};
+            {Env2, Cs2} = check_clauses_intersect(Env1, Tyss, Clauses),
+            {Env2, constraints:combine(Cs1, Cs2)};
         {fun_ty_union, Tyss, Cs1} ->
-            {VB, Cs2} = check_clauses_union(Env, Tyss, Clauses),
-            {VB, constraints:combine(Cs1, Cs2)};
+            {Env2, Cs2} = check_clauses_union(Env, Tyss, Clauses),
+            {Env2, constraints:combine(Cs1, Cs2)};
         {type_error, _} ->
             throw({type_error, Fun, type('fun'), Ty})
     end;
 
 do_type_check_expr_in(Env, ResTy, {'receive', _, Clauses}) ->
-    check_clauses(Env, [type(any)], ResTy, Clauses, capture_vars);
+    {Env1, Cs} = check_clauses(Env, [type(any)], ResTy, Clauses, capture_vars),
+    {Env1, Cs};
 do_type_check_expr_in(Env, ResTy, {'receive', _, Clauses, After, Block}) ->
-    {VarBinds1, Cs1} = check_clauses(Env, [type(any)], ResTy, Clauses, capture_vars),
-    {VarBinds2, Cs2} = type_check_expr_in(Env
-                                         ,type(integer)
-                                         ,After),
-    {VarBinds3, Cs3} = type_check_block_in(Env, ResTy, Block),
-    {union_var_binds([VarBinds1, VarBinds2, VarBinds3], Env)
-                    ,constraints:combine([Cs1, Cs2, Cs3])};
+    {Env1, Cs1} = check_clauses(Env, [type(any)], ResTy, Clauses, capture_vars),
+    {Env2, Cs2} = type_check_expr_in(Env1, type(integer), After),
+    {Env3, Cs3} = type_check_block_in(Env2, ResTy, Block),
+    {union_var_binds([Env1, Env2, Env3], Env), constraints:combine([Cs1, Cs2, Cs3])};
```

Comment:
I don't think this is correct. The variables bound in Clauses are not available in After (the timeout expression) nor the after Block. 

The vars bound in After are available in Block though. That's a bug in the legacy code. (I was unsure so I had to test it in an erl shell.)

Example:

```Erlang
receive
    V when is_atom(V) ->
        atom_to_list(V)
after (begin V=42, V+1000 end) ->     % a different V
    integer_to_list(V)
end.
```

Reply:
Hmm... So I checked again:

> The vars bound in After are available in Block though.

They don't seem to be:

```erlang
1> receive
1>     V when is_atom(V) ->
1>         atom_to_list(V)
1> after (begin V=42, V+1000 end) ->     % a different V
1>     integer_to_list(V)
1> end.
* 5:21: variable 'V' is unbound
```

Maybe it depends on the Erlang version? Though it would be quite surprising 🤔

For now, I've pushed a fix which makes each of `Clauses`, `After`, and `Block` independent.